internal/controller/controllers/csr_utils.go

```diff
@@ -0,0 +1,228 @@
+	if !hasExactUsages(csr, kubeletClientUsages) {
+		return false, errors.Errorf("CSR %s agent %s/%s: No exact match between CSR %v and required %v usages", csr.Name, agent.Namespace, agent.Name,
+			csr.Spec.Usages, kubeletClientUsages)
+	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, nodeUserPrefix) {
+		return false, errors.Errorf("CSR %s agent %s/%s: Common name %s does not have prefix %s", csr.Name, agent.Namespace, agent.Name,
+			x509cr.Subject.CommonName, nodeUserPrefix)
+	}
+
+	if !isReqFromNodeBootstrapper(csr) {
+		return false, errors.Errorf("CSR %s agent %s/%s: Not from bootstrapper", csr.Name, agent.Namespace, agent.Name)
+	}
+	return true, nil
+}
+
+// Validate that server CSR can be approved.  Server CSR can be approved after node joins the cluster  and alternative names are matched
+func validateNodeServerCSR(agent *aiv1beta1.Agent, node *corev1.Node, csr *certificatesv1.CertificateSigningRequest, x509CSR *x509.CertificateRequest) (bool, error) {
+	nodeAsking := strings.TrimPrefix(csr.Spec.Username, nodeUserPrefix)
+	if len(nodeAsking) == 0 {
+		return false, errors.Errorf("CSR %s agent %s/%s: CSR does not appear to be a node serving CSR. Empty node name after %s prefix", csr.Name, agent.Namespace, agent.Name,
+			nodeUserPrefix)
+	}
+
+	// Check groups, we need at least:
+	// - system:authenticated
+	if len(csr.Spec.Groups) < 2 || !funk.ContainsString(csr.Spec.Groups, "system:authenticated") {
+		return false, errors.Errorf("CSR %s agent %s/%s: %v is too small or not contains %q", csr.Name, agent.Namespace, agent.Name,
+			csr.Spec.Groups, "system:authenticated")
+	}
+
+	// Check usages, we need only:
+	// - digital signature
+	// - key encipherment
+	// - server auth
+	serverUsages := []certificatesv1.KeyUsage{
+		certificatesv1.UsageDigitalSignature,
+		certificatesv1.UsageKeyEncipherment,
+		certificatesv1.UsageServerAuth,
+	}
+
+	if !hasExactUsages(csr, serverUsages) {
+		return false, errors.Errorf("CSR %s agent %s/%s: No exact match between CSR %v and required %v usages", csr.Name, agent.Namespace, agent.Name,
+			csr.Spec.Usages, serverUsages)
+	}
+	if !funk.ContainsString(x509CSR.Subject.Organization, nodeGroup) {
+		return false, errors.Errorf("CSR %s agent %s/%s: Organization %v doesn't include %s", csr.Name, agent.Namespace, agent.Name,
+			x509CSR.Subject.Organization, nodeGroup)
+	}
+	if x509CSR.Subject.CommonName != csr.Spec.Username {
+		return false, errors.Errorf("CSR %s agent %s/%s: Mismatched CommonName %s != %s for CSR %s", csr.Name, agent.Namespace, agent.Name,
+			x509CSR.Subject.CommonName, csr.Spec.Username, csr.Name)
+	}
+	agentDNSNames := append(getNodeDNSNames(node), getAgentHostname(agent))
+	for _, dnsName := range x509CSR.DNSNames {
+		if !funk.ContainsString(agentDNSNames, dnsName) {
+			return false, errors.Errorf("CSR %s agent %s/%s: DNS name %s missing from available DNS names %v", csr.Name, agent.Namespace, agent.Name,
+				dnsName, agentDNSNames)
+		}
+	}
+	agentIPs := getAgentIPs(agent)
```

Comment:
As in the previous comment, we should probably using the node's IP here as that's the info that should have been used for the CSR.

There's no agent in the spoke cluster, so probably we should not be matching the CSR with the agent itself

Reply:
OK